src/QirRuntime/public/IQuantumApi.hpp

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <complex>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    struct IQuantumApi
+    {
+        virtual ~IQuantumApi() {}
+
+        // TODO: Only some simulators 'know' the state -- is it OK to have this method on IQuantumApi?
+        // The callback will be invoked on each basis vector (in the standard computational basis) in little-endian
+        // order until it returns `false` or the state is fully dumped.
+        typedef bool (*TGetStateCallback)(size_t /*basis vector*/, double /* amplitude Re*/, double /* amplitude Im*/);
+        virtual void GetState(TGetStateCallback callback) = 0;

```

Comment:
To answer the TODO: no, it's not OK to have this method in the base interface. It's not just "only some": it's basically one (full-state simulator) that has a state that is accessible in this way, compared to many (CHP, density matrix, resource counter, ...) that don't.

Reply:
Moved it to IDiagnostics